processing/src/main/java/org/apache/druid/frame/write/cast/ObjectToExprEvalColumnValueSelector.java

```diff
@@ -0,0 +1,110 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.write.cast;
+
+import org.apache.druid.math.expr.ExprEval;
+import org.apache.druid.math.expr.ExpressionType;
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.RowIdSupplier;
+import org.apache.druid.segment.column.ColumnType;
+
+import javax.annotation.Nullable;
+
+/**
+ * Wraps a {@link ColumnValueSelector}, calls {@link ColumnValueSelector#getObject()}, and casts that value using
+ * {@link ExprEval}. The object is interpreted using {@link ExprEval#ofType}.
+ */
+public class ObjectToExprEvalColumnValueSelector implements ColumnValueSelector<Object>
+{
+  private final ColumnValueSelector<?> selector;
+  @Nullable
+  private final ExpressionType selectorType;
+  private final ExpressionType desiredType;
+  @Nullable
+  private final RowIdSupplier rowIdSupplier;
+
+  public ObjectToExprEvalColumnValueSelector(
+      final ColumnValueSelector<?> selector,
+      @Nullable final ColumnType selectorType,
+      final ColumnType desiredType,
+      @Nullable final RowIdSupplier rowIdSupplier
+  )
+  {
+    this.selector = selector;
+    this.selectorType = ExpressionType.fromColumnType(selectorType);
+    this.desiredType = ExpressionType.fromColumnType(desiredType);
+    this.rowIdSupplier = rowIdSupplier;
+  }
+
+  @Override
+  public double getDouble()
+  {
+    return eval().asDouble();
+  }
+
+  @Override
+  public float getFloat()
+  {
+    return (float) eval().asDouble();
+  }
+
+  @Override
+  public long getLong()
+  {
+    return eval().asLong();
+  }
+
+  @Override
+  public boolean isNull()
+  {
+    return eval().isNumericNull();
+  }
+
+  @Nullable
+  @Override
+  public Object getObject()
+  {
+    return eval().value();
+  }
+
+  @Override
+  public Class<Object> classOfObject()
+  {
+    return Object.class;
+  }
+
+  @Override
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    inspector.visit("selector", selector);
+    inspector.visit("rowIdSupplier", rowIdSupplier);
+  }
+
+  private ExprEval<?> eval()
+  {
+    final Object obj = selector.getObject();
+    if (obj == null) {
+      return ExprEval.of(null);
```

Comment:
nit: could use `ExprEval.ofType(selectorType, null)` so it doesn't become a `ExpressionType.STRING`

Reply:
It still does become a string if we do that, b/c of how the code flow works through `ExprEval.ofType`. Maybe it shouldn't? But it does.

Reply:
if type is null it becomes a string, but if the value is null it is an `ExprEval` of that type with a null value